mesonbuild/dependencies/base.py

```diff
@@ -1078,21 +1093,23 @@ def description(self):
         '''Human friendly description of the command'''
         return ' '.join(self.command)
 
-    @staticmethod
-    def from_bin_list(bins, name):
-        if name not in bins:
+    @classmethod
+    def from_bin_list(cls, bt: BinaryTable, name):
```

Comment:
Mmm from my quick checks they seem to do nothing (other than be name-resolved)  in the min supported version of python. Can you measure this? I quite like them for documentation now, and look forward to python 3.7 and beyond making such refactors as I've been doing much easier.

Reply:
I can't find it now, but the reason pep 563 was written (the pep that adds `__future__.annotations` in 3.7) is because at module import time all type annotations have to be resolved, ie, instances of class types are instantiated. It is non-free, it can have a very real cost at startup time.

Reply:
I dove into https://www.python.org/dev/peps/pep-0563/ a bit. The cost they refer to is just in the computation of the annotations; the annotations are never checked. Every annotation i write is just a class. The class exists and must be evaluated, so this should be basically free. It's only stuff like:
```python
_: MyGeneric[MyOtherGeneric[..], MyOtherGeneric[..]]
```
that has anything meaningful to evaluate. And even then it can't be too expensive. It's basically an O(n) with some sublinear scope lookups.

I strongly urge not worrying about this unless we have benchmarks to prove its a problem.

Reply:
I can tell you from experience type annotations before `__future__.annotations` are not free. I bumped a project I had to 3.7 and added the future import and my startup time dropped from several seconds to instantaneous. Apart from that these are useless unless a) we adopt them project wide and b) we start using a tool like mypy to ensure that they're actually right. It also feels sneaky to start using them in a completely unrelated PR.

Please remove them from this PR. Send a separate PR and add them to the whole project if you want to propose using them.

Reply:
AFAICR we do have annotations somewhere in the code base but my grep-fu is failing me. The discussion at the time was very similar to this one, but we chose to allow them since they do work from 3.5 onwards, which is the oldest supported version. Eventually we want to have annotations, though, and adding them incrementally is fine.

That being said if there is a noticeable performance impact then we should be careful. Meson runs fairly slow already.